Lake/Build/Actions.lean

```diff
@@ -43,16 +43,22 @@ def compileOlean (leanFile oleanFile : FilePath)
 def compileOleanAndC (leanFile oleanFile cFile : FilePath)
 (oleanPath : SearchPath := []) (rootDir : FilePath := ".")
 (leanArgs : Array String := #[]) (lean : FilePath := "lean")
+(precompiledPath : SearchPath := ∅)
 : BuildM PUnit := do
   createParentDirs cFile
   createParentDirs oleanFile
+  let mut env := #[("LEAN_PATH", some oleanPath.toString)]
+  if !precompiledPath.isEmpty && System.Platform.isWindows then
+    let path := (← IO.getEnv "PATH").getD ""
+    -- set up search path for locating dependent DLLs
```

Comment:
Presumably this shouldn't be just `precompiledPath`, but `soPath` and also include DLL directories from dependencies.

Reply:
What do you mean by `soPath`? I haven't looked into package dependencies yet, true. Frankly I'm fine for now if it works on mathlib4 :smile: (esp. if it works well enough so that we can benchmark and detect important bottlenecks).

Reply:
@Kha I'm actually a little confused about how this works, why do you even need to change the PATH? Why is passing the full path to --load-dynlib not sufficient?

Reply:
The full path works, but it's that library's *dependencies* that are searched by name in PATH. We only pass the direct dependencies' libraries to `lean`, which ensures the cmdline length is reasonably bounded.

Reply:
@Kha Why would command line length be a concern? If passing the entire dependency closure works, that seems like a better solution to me than altering `PATH`. Furthermore, from the code in `recBuildModulePrecompTargetWithLocalImports`, it looks like that such an approach already necessary on Windows, so keeping it consistent across platforms seems like another benefit.

**Note:** I am not asking you change the PR, I just want to verify that such a solution is feasible for when I implement this feature myself (hopefully) this week,

Reply:
Command line lengths are *always* a concern for practically unbounded input. For the linker we can always use response files, which we have yet to invent for Lean. But it will probably take a while for dependency trees to get large enough for this to blow up, so passing the entire dependency tree sounds good to me.

I don't understand the "consistent across platforms" comment, other platforms should not unnecessarily be burdened with Windows limitations.

Reply:
Maybe this is due to some experience I lack, but I still don't see why lengths are a problem. Is there some major performance or security issue with lengths that go even into the multiples of MBs range? It seems like that still wouldn't be a major burden to memory or parse time (though I can see why one might not want to log it).

For the cross-platform code, sure, special casing an operating system when there is clear advantage to for doing so makes sense. However, if multiple approaches are reasonable and one is cross-platform, I would go with it. I think simpler code is a virtue and, if it doesn't have a significant cost, should be preferred.

Reply:
Ah, now I understand the confusion! Command line lengths are unfortunately quite limited in practice: https://devblogs.microsoft.com/oldnewthing/20031210-00/?p=41553. But I think we should fall under the 32K case, so we should be good for a good while. Still, people (especially people with unusual build setups like in Nixpkgs and GHC that create an unusual number of dependent libraries) do run into these limits on all platforms from time to time, which is why they invented response file to hide their humongous command-line arguments in.

Reply:
Thanks! I learn something new everyday. I am somewhat surprised that such a limitation hasn't been lifted yet (either on Windows or Unix), but I imagine there are complexities there I do not realize.

We should fall under 32K for now. However, if we assume an average of 50-80 bytes per argument, that only allows somewhere between 400-640 modules to be loaded by Lean or linked by the linker, so we will likely hit that cap rather soon. I can easily see mathlib surpassing it relatively quickly (and definitely by the time the port is finished). Thus, response files for CLIs (both Lean and the linker) seem like an important feature to consider soon.